### ISSUE TITLE
grad allocation in simple_bind

### ISSUE THREAD

**Issue by pluskid** [dmlc/mxnet#267]
I'm reading the python front end for `simple_bind`, we have 
```python
if not (name.endswith('data') or name.endswith('label')):
    grad_ndarrays[name] = zeros(shape, ctx)
```
It seems a little bit arbitrary to detect a data symbol by its name ending with `data` or `label`. At least we should document this somewhere. But otherwise, I think we could have a more explicit way:

Allow users attach *attributes* to symbols, e.g. marking some symbol as `data`, or as `frozen`. Consider at least the following (quite common) scenarios:
* The data symbol might not be named with a `data` ending. Maybe, for example, in recurrent models, they might be named `data_t0`, `data_t1`, `data_t2`, ...  etc.
* The user want to fix some of the layers while only train other layers. This is quite common in layer-wise pre-training, as well as training complicated multi-modal architectures.
* The user want to compute gradient for the data and back-propagate into the data while fixing the weights. This is used to do visualization of networks or generate adversarial examples for networks. ~~Or maybe someone want to use this to generate an ideal test set on which a given model performs really well.~~

While the users can always use the raw `bind` to achieve full control. Do you think it would be good to have a `not_so_simple_bind` that does a bit more than `simple_bind`?

**Comment by tqchen** [comment]
I like the idea of attributes to symbols. This will requires a bit change in the c++ representation of symbol, though, but can make the encoding of name and type a bit more explicit.

**Comment by mli** [comment]
the current solution for multi-input or multi-label is `data_0`, ..., `data_n`, .and `label_0`, ... `label_m`. and ask the user to provide an iterator returning the data in the same order.

**Comment by pluskid** [comment]
@mli Is there any example? I saw @antinucleon's LSTM example is build from scratch using low level symbol API so he knows exactly which symbol is data and which is label.

**Comment by antinucleon** [comment]
data seq is same to ```symbol.list_arguments()``` and outputs seq is same to ```symbol.list_outputs```. So simply we can make a dict ```dict(zip(sym.list_arguments(), exe.arg_arrays))``` to map from name to NDArrary

**Comment by pluskid** [comment]
@antinucleon I see, thanks! The LSTM code example is actually very clear.

**Comment by pluskid** [comment]
Attributes is still interesting but the priority should be pretty low. I'm closing this issue for now as currently `simple_bind` already works with arbitrary naming convention as long as the data names and shapes from the data iter are accurate.
